consensus/consensus_test.go

```diff
@@ -0,0 +1,1023 @@
+		i++
+	}
+
+}
+
+func TestSenderSendsTwoMessagesWithDelay(t *testing.T) {
+	numOfNodes := 20
+
+	var dsInstances []Algorithm
+
+	cfg := config.DefaultConfig()
+	cfg.NodesPerLayer = int32(numOfNodes)
+	cfg.NumOfAdverseries = int32(numOfNodes / 2)
+	timer := MyTimer{}
+
+	mockNetwork := NewMockNetwork(0)
+	quit := mockNetwork.routeMessages()
+	zeroBuf := make([]byte, 0)
+
+	nodeIds, nodes := createNodesAndIds(numOfNodes)
+
+	for i := 0; i < numOfNodes -1; i++ {
+		ds := initDSC(mockNetwork, nodeIds, nodes[i], cfg, timer)
+		dsInstances = append(dsInstances, ds)
+	}
+
+	for i := 0; i < numOfNodes-1; i++ {
+		x := messageData(zeroBuf)
+		//start instance blocks until protocol is finished
+		go dsInstances[i].StartInstance(x)
+	}
+
+	node := nodes[numOfNodes -1]
+	nodeMock := mockNetwork.createNewNodeNetMock(node.publicKey.String())
+	dsc, _ := createDSC(cfg, timer, nodeIds, nodeMock, node.publicKey, node.privateKey)
+	mal := MaliciousDelayedSecondMessageSender{*dsc,}
+
+	//mal, _ := initMaliciousDSC(mockNetwork, nodeIds, nodes[numOfNodes -1], cfg, timer)
+
+	arr := []byte{'l', 'o', 'l'}
+	msg := messageData(arr)
+	msg2 := messageData([]byte("lol2"))
+	mal.StartInstance(msg, msg2)
+
+	log.Info("closing network")
+	quit <- struct{}{}
+	i := 0
+	for _, dsc := range dsInstances {
+		validateDSC(t, dsc, arr, nodeIds[i], nodeIds[numOfNodes-1], false)
+		i++
+	}
+
+}
+
+func initDolevStrongConsensus(numOfNodes int) *DolevStrongConsensus{
+	cfg := config.DefaultConfig()
+	cfg.NodesPerLayer = int32(numOfNodes)
+	cfg.NumOfAdverseries = int32(numOfNodes / 2)
+	timer := MyTimer{}
+
+	mockNetwork := NewMockNetwork(0)
+	//quit := mockNetwork.routeMessages()
+	node := NewNode()
+
+	nodeMock := mockNetwork.createNewNodeNetMock(node.publicKey.String())
+	dolevStrong, _ := createDSC(cfg,timer,nil,nodeMock,node.publicKey,node.privateKey)
+	return dolevStrong
+}
+
+func TestInvalidPublicKeyOfInitiatorMessage(t *testing.T){
+	numOfNodes := 10
+	dolevStrong := initDolevStrongConsensus(numOfNodes)
+
+	arr := messageData([]byte("anton"))
+	msg := createMessage(arr ,dolevStrong)
+
+	// flip a byte in the pub key to fail
+	msg.Msg.AuthPubKey[0] = ^msg.Msg.AuthPubKey[0]
+
+	bt, _ := proto.Marshal(&msg)
+	ms := messageData(bt)
+	err := dolevStrong.handleMessage(ms)
+
+	assert.Error(t, err)
+}
+
+func TestInvalidSignature(t *testing.T){
+	numOfNodes := 10
+	dolevStrong := initDolevStrongConsensus(numOfNodes)
+	msg := createMessage(messageData([]byte("anton")), dolevStrong)
+	dolevStrong.signMessageAndAppend(&msg)
+	msg.Msg.Data[0] = ^msg.Msg.Data[0]
+	instance := DolevStrongInstance{
+		ds:dolevStrong,
+	}
+	_, err := instance.findAndValidateSignatures(&msg)
+	assert.Error(t,err)
+}
+
+func validateDSC(t *testing.T, dsc Algorithm, expected []byte, node string, senderInstance string, shouldExist bool) {
+	outputs := dsc.GetOtherInstancesOutput()
+	for key, out := range outputs {
+		if key == senderInstance && node != senderInstance {
```

Comment:
I don't understand this if. Why are you only validating that the sender has the correct output? What about the rest of the receivers?